### ISSUE TITLE
Option to do custom coercions/type casting?

### ISSUE THREAD

**Issue by eXigentCoder** [loris/api-query-params#80]
Firstly, thanks for a really useful library, does 99% of what I need to do.

I'm looking for a way to coerce an specific field in the filter result to a mongodb.ObjectId before I run the query against the db, is there currently a way to do this in the library?

So for example a way to do it manually at the moment would be:
```javascript
'use strict';
var mongo = require('mongodb');
var aqp = require('api-query-params').default;

var queryString = '?ownderId:5829ba2b7a3a7a08307483a9';
var parsedQueryString = aqp(queryString);
parsedQueryString.ownderId = mongo.ObjectId(parsedQueryString.ownderId);
mongo.db.collection('items').findOne(parsedQueryString.filter, dataRetrieved);
```

In the above example it's fairly trivial, but it would be nice to have an extensible way of doing coercion before the query goes into the db. The way I see it there are a few options if this functionality doesn't already exist:
# Custom functions
Allow setting up of custom "functions" in the query string before parsing, in a similar manner to: 

```javascript
aqp('key1=string(10)&key2=date(2016)&key3=string(null)');
```

For example

```javascript
aqp.addFunction('mongoId', function(stringValue, keyName, fullQueryString){
    return mongo.ObjectId(stringValue);
});
aqp('?ownderId:mongoId(5829ba2b7a3a7a08307483a9)');
```

# Key Name Lookup
Use the key name in the query string to do type casting

```javascript
aqp.coerce(['ownderId'], function(stringValue, keyName, fullQueryString){
    return mongo.ObjectId(stringValue);
});
aqp('?ownderId:5829ba2b7a3a7a08307483a9');
```

Apologies if there is already a way to do this.

**Comment by loris** [comment]
Glad to head this library is fitting your need! 
What you ask is not implemented at the moment, but what you proposed looks great and would be a nice addition. I will take some time off today or tomorrow to make a PR.
IMO, both options (custom functions and key name lookup) are useful, but I would change the "key name lookup" to let the user specify a parsing function (existing one, or a custom one)
Additionally, using `coerce`/`addFunction` methods would require a breaking change: making `aqp` a constructor. I’d rather keep this for a major release, and just use the options params for the time being:

```
const options = {
  parsers: {
    id: value => mongo.ObjectId(value),
    float: value => parseFloat(value),
  }
};
```

This way you could explicitly use the new parsers in the query:
```aqp('?ownerId=id(5829ba2b7a3a7a08307483a9)&balance=float(10)', options);```

Or implicitly by specifying a mapping between key and parser:

```
const options = {
  parsers: {
    id: value => mongo.ObjectId(value),
    float: value => parseFloat(value),
  },
  mapParserToKeys: {
    ownerId: 'id',
    balance: 'float',
  }
};
aqp('?ownerId=5829ba2b7a3a7a08307483a9&balance=10', options);
```
What do you think?

**Comment by eXigentCoder** [comment]
Thanks for taking the time to look at this!
The solution you propose above makes a lot of sense. A breaking change that would make `aqp` into a constructor would probably be a lot of work since all the examples and `README` doc would need to be updated as well.

Ultimately I think the second example where you can set parsers and maps is really useful since it can accommodate a lot of different peoples use-cases.

**Comment by loris** [comment]
@eXigentCoder Just released 4.2 with both options (`casters` and `castParams`, see README.md or test cases for more information), hope this helps :)

**Comment by eXigentCoder** [comment]
Just finished updating my code to use the new version of the library, and it worked perfectly, very awesome, thanks a ton!
